### PR TITLE
Add multiple path support

### LINKED ISSUE #95: Accept more than one project
Currently, `cargo sweep` accepts one optional and positional argument as path to the project to `sweep`.

It would be good to have it accepting multiple paths.

```
cargo sweep [OPTS] project1 project2 project3
```

`cargo-sweep` is already capable of sweeping multiple projects (see `--recursive` implementation), so this _should_ be easy to implement.

### PR DESCRIPTION
**Pull request by social-anthrax**
Closes #95.

Adds support to process multiple paths.
Timestamps are now read from paths[0] instead of the only passed in path (This is probably not expected behaviour).

## Breaking Changes
- `cargo sweep --toolchains 1 2 3` no longer causes the expected error as 2 and 3 are read as paths.

### CODE REVIEW

File: tests/integration.rs [hunk 1]
```diff
@@ -518,7 +518,7 @@ fn multiple_paths() -> TestResult {
     assert_eq!(final_build_size, size_after_dry_run_clean);
 
     // Run a proper cargo-sweep ("clean") in the target directories
-    let mut args = vec!["-r", "--time", "0"];
+    let mut args = vec!["--time", "0"];
```
Comment (jyn514) [inline tests/integration.rs:521]: Why this change?

File: src/cli.rs [hunk 2]
```diff
@@ -37,7 +37,7 @@ pub enum SweepCommand {
 )]
 pub struct Args {
     /// Path to check
-    pub path: Option<PathBuf>,
+    pub path: Option<Vec<PathBuf>>,
```
Comment (jyn514) [inline src/cli.rs:40]: Can this just be `Vec<PathBuf>` instead?

File: tests/integration.rs (hunk 1 above)
Comment (social-anthrax) [reply tests/integration.rs:521]: We only want to test the functionality for multiple paths, setting both the multiple paths and recursive could cause an error in either one so I just wanted to minimise the test case.

File: src/main.rs [hunk 3]
```diff
@@ -138,30 +137,50 @@ fn main() -> anyhow::Result<()> {
...
+    };
+    // FIXME: Change to write to every passed in path instead of just the first one
+    // if let Criterion::Stamp = criterion {
+    //     return paths.iter().try_for_each(|path| {
+    //         debug!("Writing timestamp file in: {:?}", path);
+    //         Timestamp::new()
+    //             .store(path.as_path())
+    //             .context("Failed to write timestamp file")
+    //     });
+    // }
```
Comment (jyn514) [inline src/main.rs:162]: If you want to keep this FIXME, please move it to line 144, but still delete this dead code.
```suggestion
```

File: src/main.rs [hunk 4]
```diff
@@ -206,22 +225,23 @@ fn main() -> anyhow::Result<()> {
...
-            let ts =
-                Timestamp::load(path.as_path(), dry_run).expect("Failed to load timestamp file");
+            let ts = Timestamp::load(paths[0].as_path(), dry_run)
+                .expect("Failed to load timestamp file");
             Duration::from(ts)
         } else if let Criterion::Time(days_to_keep) = criterion {
             Duration::from_secs(days_to_keep * 24 * 3600)
         } else {
+            dbg!(criterion);
             unreachable!();
```
Comment (jyn514) [inline src/main.rs:235]:
```suggestion
            unreachable!("unknown criteria {:?}", criterion);
```

File: src/cli.rs [hunk 5]
```diff
@@ -158,6 +159,5 @@ mod tests {
             ..Args::default()
         };
         assert_eq!(args, parse("cargo sweep --toolchains 1,2,3").unwrap());
-        assert!(parse("cargo sweep --toolchains 1 2 3").is_err());
```
Comment (jyn514) [inline src/cli.rs:161]: Can you add a test that the equivalent with "1 2 3" as a single argument (instead of using split_whitespace) is still an error?

File: tests/integration.rs [hunk 6]
```diff
@@ -457,3 +458,122 @@ fn recursive_multiple_root_workspaces() -> TestResult {
 
     Ok(())
 }
+
+/// This test follows the logic of the recursive multiple root test, however, instead of recursing it passes each workspace individually.
+
+#[test]
+fn multiple_paths() -> TestResult {
```
Comment (jyn514) [inline tests/integration.rs:465]: This test still passes even without your changes. Not sure what's going wrong - maybe you're only asserting that the total size has decreased, instead of that both projects are being cleaned?

**Review by jyn514 — CHANGES_REQUESTED** [review]
Thanks, this looks good! Just a couple nits :)

File: tests/integration.rs (hunk 6 above)
Comment (social-anthrax) [reply tests/integration.rs:465]: Removing my change locally makes this fail so I'm also not sure what's going on 😅

Comment (jyn514) [reply tests/integration.rs:465]: ah, I was testing this wrong, sorry - yes, I can also reproduce this failing without your change :) sorry for the noise.

Comment (social-anthrax) [reply tests/integration.rs:465]: No worries!

File: src/cli.rs (hunk 2 above)
Comment (social-anthrax) [reply src/cli.rs:40]: Having it as an option makes defaulting to the current invocation path easier imo, happy to change it though!

Comment (social-anthrax) [reply src/cli.rs:40]: Easy to change it to this if you'd prefer
```rust
    let paths = match args.path.len() {
        0 => vec![env::current_dir().expect("Failed to get current directory")],
        _ => args.path,
    };
```

Comment (jyn514) [reply src/cli.rs:40]: When is it possible to have `Some(vec![])` here? If the answer is "never" then I prefer the `match` you just posted, yeah.

File: src/cli.rs (hunk 5 above)
Comment (jyn514) [reply src/cli.rs:161]: Oh right, this is https://github.com/rust-lang/rustup/issues/3168 again. Ok, no need for the new test.

File: src/main.rs (hunk 4 above)
Comment (social-anthrax) [reply src/main.rs]: Fixed

File: src/main.rs (hunk 3 above)
Comment (social-anthrax) [reply src/main.rs]: Kept the FIXME, removed the deadcode

File: src/cli.rs (hunk 2 above)
Comment (social-anthrax) [reply src/cli.rs]: Added

**Review by jyn514 — APPROVED** [review]

**Comment by jyn514** [comment]
Thanks!
